File: Medium/2352. Equal Row and Column Pairs/EqualRowAndColumnPairs.py
'''
LeetCode Medium: 2352. Equal Row and Column Pairs
https://leetcode.com/problems/equal-row-and-column-pairs/

Check EqualRowAndColumnPairs.cpp for brief solution explanation.
'''
class Solution:
    def equalPairs(self, grid: List[List[int]]) -> int:
        N = len(grid)
        SEED, MOD = 131, int(1e9+7)
        row_hash = [0 for i in range(N)]
        for r in range(N):
            for i in range(N):
                row_hash[r] = ((row_hash[r]*SEED)%MOD+grid[r][i])%MOD
        col_hash = [0 for i in range(N)]
        for c in range(N):
            for i in range(N):
                col_hash[c] = ((col_hash[c]*SEED)%MOD+grid[i][c])%MOD
        ans = 0
        for r in range(N):
            for c in range(N):
                if row_hash[r]==col_hash[c]:
                    ans+=1
        return ans

File: Medium/2352. Equal Row and Column Pairs/EqualRowAndColumnPairs.java
/*
LeetCode Medium: 2352. Equal Row and Column Pairs
https://leetcode.com/problems/equal-row-and-column-pairs/

Check EqualRowAndColumnPairs.cpp for brief solution explanation.
*/
class Solution {
    public int equalPairs(int[][] grid) {
        int N = grid.length;
        int SEED = 131, MOD = (int)1e9+7;
        long[]row_hash = new long[N];
        for(int r = 0;r<N;r++){
            for(int i = 0;i<N;i++){
                row_hash[r] = ((row_hash[r]*SEED)%MOD+grid[r][i])%MOD;
            }
        }
        long[]col_hash = new long[N];
        for(int c = 0;c<N;c++){
            for(int i = 0;i<N;i++){
                col_hash[c] = ((col_hash[c]*SEED)%MOD+grid[i][c])%MOD;
            }
        }
        int ans = 0;
        for(int r = 0;r<N;r++){
            for(int c = 0;c<N;c++){
                if(row_hash[r]==col_hash[c]){
                    ans++;
                }
            }
        }
        return ans;
    }
}